### ISSUE TITLE
Rule proposal: `string-content`

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#327]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2480%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/327)

### Good 👍 

```js
const description = 'Add text to user’s comments'
```

### Not good 👎 

```js
const description = 'Add text to user\'s comments'
```

Prior art: https://github.com/jmont/eslint-plugin-smart-quotes

I tried that one but it doesn't have a fixer and it has many false positives. Also strings could be CSS selectors so`[alt=""]` needs to stay as is.

---

_Originally posted by @sindresorhus in https://github.com/sindresorhus/refined-github/pull/2167_:

```js
'use strict';

const create = context => {
	return {
		Literal: node => {
          	const {value} = node;
     
			if (typeof value !== 'string') {
				return;
			}
          
        	const fixedValue = value.replace(/'/, '‘');

			if (fixedValue === value) {
				return;
			}

			context.report({
				node,
				message: 'Use other quote kind',
				fix: fixer => fixer.replaceTextRange([node.range[0] + 1, node.range[1] - 1], fixedValue)
			});
		}
	};
};

module.exports = {
	create,
	meta: {
		type: 'problem',
		fixable: 'code'
	}
};
```

https://astexplorer.net/#/gist/3a67df351a57608ae38a6b55102376d6/c8425e326ddce75bf7242295c8818134ef2d2e16

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $80.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)

### Submitted pull Requests
- [#496 Add `string-content` rule](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/496)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by MrHen** [comment]
I'm not sure how a lint rule would detect when it is appropriate to use one over the other. If you have a string such as `Sally said, "Bob said, 'I am hungry'."` you cannot flag both single quotes for conversion. The strings themselves could be split up across multiple literals so you couldn't check the content of any particular literal to know which is appropriate.

There are similar potential rules with other shorthand strings:

* `"` versus `“` and `”` (which is locale specific)
* `...` versus `…`
* `->` versus `→` (or a plethora of other arrows)

I think the ellipsis would be the easiest to detect because exactly three periods in a row is pretty indicative of an ellipsis. But I'm wary of trying to challenge the content in strings -- I think linting is more suited for checking code syntax and style than English syntax and style.

I suppose we could make this a generic string content scanner that took configured regex search / replace:

`
"unicorn/string-content": ["warn", {
  "patterns": [
    {
      "match": "'",
      "suggest": "’",
      "fix": false
    },
    {
      "match": "...",
      "suggest": "…",
      "fix": true
    }
  ]
}],
`

Or some such.

**Comment by bfred-it** [comment]
My fear is that it might cause some false positive, that’s why it should probably be limited to single apostrophes as exclusively to avoid to escape them, rather than as a generic “typography” rule. 

The replacer would be great too, as a generic rule

**Comment by bfred-it** [comment]
To test the applicability of this rule in your projects, try running that plugin and looking for situations where `\'` needs to stay like that.

```sh
npm i eslint-plugin-smart-quotes -g

eslint --no-eslintrc --parser-options=ecmaVersion:8 --plugin smart-quotes --rule 'smart-quotes/smart-quotes: 2' '**/*.js' | grep "\\\'" -C 1
```

The plugin will show a lot of errors but just look for single apostrophes

**Comment by sindresorhus** [comment]
I like the idea of a generic `string-content` rule. Let's go with that.

I think `match` could be even more flexible by accepting: `string | RegExp | Array<string | RegExp>` (?)

**Comment by sindresorhus** [comment]
I would use the rule for important things:

```json
"unicorn/string-content": ["warn", {
  "patterns": [
    {
      "match": "unicorn",
      "suggest": "🦄",
      "fix": true
    }
  ]
}]
```

**Comment by otofu-square** [comment]
I'd like to work on this issue :eyes:

**Comment by fregante** [comment]
It's all yours :)

**Comment by sindresorhus** [comment]
Just make sure you read [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/new-rule.md) thoroughly and have an understanding of AST and ESLint rules :)

**Comment by fisker** [comment]
@sindresorhus I have a problem on report message, if we treat match as `RegExp`

```js
{
	match: 'https?:\\\/\\\/www\\\\.github\\\\.com',
    suggest: 'github.com'
}
```

How do we report? 

1. Message: ``Prefer `github.com`.``
2. Add `pattern.message`, and default to ``Prefer `github.com`.``
3. Report the pattern, Message ``Prefer `github.com` over `/https\?:\/\/www\\.github\\.com/gu`.``
4. Don't support `RegExp`

Another problem, this rule original idea to fix `'` to `’`, so default to 

```js
[
	{
	  match: '\\'',
	  suggest: '‘',
	}
]
```

`patterns` should override or append? or more option?

**Comment by sindresorhus** [comment]
> How do we report?

I would go with both 2 and 3. Default to 3, but allow overriding the message.

> patterns should override or append? or more option?

We could make the key be the match:

```js
{
	'\\'': {
		suggest: '‘'
	}
}
```

And allow setting it to `false`. Then the user could either override or disable with:

```js
{
	'\\'': false
}
```

**Comment by sindresorhus** [comment]
Rewarding this issue is blocked by a dumb IssueHunt bug: https://spectrum.chat/issuehunt/general/error-when-accessing-an-issue-page~edd3be59-9228-472f-ae32-ce888e56bc1c?m=MTU4MzcyNjY3MjIxOQ==

**Comment by isghe** [comment]
It's a dangerous rule https://github.com/xojs/xo/pull/439#issuecomment-598751860
